akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala

```diff
@@ -359,127 +359,129 @@ object ShardCoordinator {
   var unAckedHostShards = Map.empty[ShardId, Cancellable]
   // regions that have requested handoff, for graceful shutdown
   var gracefulShutdownInProgress = Set.empty[ActorRef]
   var aliveRegions = Set.empty[ActorRef]
-  var persistCount = 0
 
   import context.dispatcher
   val rebalanceTask = context.system.scheduler.schedule(rebalanceInterval, rebalanceInterval, self, RebalanceTick)
 
   Cluster(context.system).subscribe(self, ClusterShuttingDown.getClass)
 
+  implicit val node = Cluster(context.system)
+  val CoordinatorStateKey = LWWRegisterKey[State](s"${typeName}CoordinatorState")
+
+  getState()
+
   override def postStop(): Unit = {
     super.postStop()
     rebalanceTask.cancel()
     Cluster(context.system).unsubscribe(self)
   }
 
-  override def receiveRecover: Receive = {
-    case evt: DomainEvent ⇒
-      log.debug("receiveRecover {}", evt)
+  override def receive: Receive = waitingForState
+
+  def waitingForState: Receive = {
+    case g @ GetSuccess(CoordinatorStateKey, _) ⇒
+      state = g.get(CoordinatorStateKey).value
+      state.regionProxies.foreach(context.watch)
+      state.regions.foreach { case (a, _) ⇒ context.watch(a) }
+      state.shards.foreach { case (a, r) ⇒ sendHostShardMsg(a, r) }
+      allocateShardHomes()
+      activate()
+
+    case GetFailure(CoordinatorStateKey, _) ⇒
+      log.error(
+        "The ShardCoordinator was unable to get an initial state within 'waiting-for-state-timeout' (was retrying): {} millis",
+        waitingForStateTimeout.toMillis)
+      getState()
+
+    case NotFound(CoordinatorStateKey, _) ⇒ activate()
+  }
+
+  def active: Receive = {
+    case UpdateSuccess(CoordinatorStateKey, Some(evt: DomainEvent)) ⇒
+      log.debug("The coordinator state was successfully updated with {}", evt)
       evt match {
         case ShardRegionRegistered(region) ⇒
-          persistentState = persistentState.updated(evt)
-        case ShardRegionProxyRegistered(proxy) ⇒
-          persistentState = persistentState.updated(evt)
-        case ShardRegionTerminated(region) ⇒
-          if (persistentState.regions.contains(region))
-            persistentState = persistentState.updated(evt)
-          else {
-            log.debug("ShardRegionTerminated, but region {} was not registered. This inconsistency is due to that " +
-              " some stored ActorRef in Akka v2.3.0 and v2.3.1 did not contain full address information. It will be " +
-              "removed by later watch.", region)
-          }
-        case ShardRegionProxyTerminated(proxy) ⇒
-          if (persistentState.regionProxies.contains(proxy))
-            persistentState = persistentState.updated(evt)
-        case ShardHomeAllocated(shard, region) ⇒
-          persistentState = persistentState.updated(evt)
-        case _: ShardHomeDeallocated ⇒
-          persistentState = persistentState.updated(evt)
+          val firstRegion = state.regions.isEmpty
+
+          state = state.updated(evt)
```

Comment:
The state was already updated in the [update](https://github.com/akka/akka/pull/17871/files#diff-e3b09e26a65bebe3b08cd0b802fde5e8R601) method.

The order of the events applied to the state is important. The original order can be changed if we apply the event here, because of the [retry](https://github.com/akka/akka/pull/17871/files#diff-e3b09e26a65bebe3b08cd0b802fde5e8R445) and that UpdateSuccess messages are not guaranteed to be received in the same order as corresponding Update messages were emitted.

I see two options.
1) Apply the even and set the local state before sending Update to the replicator. Some of the code here must also be moved to that point. No retry. We rely on the gossip replication in case WriteMajority timeout.
2) We apply the event and set the local state here instead of doing it before sending Update to the replicator. We retry until UpdateSuccess. We stash all incoming messages until we receive UpdateSuccess, i.e. we perform only one update at a time.

The second alternative is more safe in case of fail over to another coordinator instance.